pika/adapters/select_connection.py

```diff
@@ -81,105 +93,40 @@ class IOLoop(object):
-        :rtype: str
-
-        """
-        if not self.poller:
-            time.sleep(deadline)
-            return callback_method()
-        return self.poller.add_timeout(deadline, callback_method)
-
-    @property
-    def poller_type(self):
-        """Return the type of poller.
-
-        :rtype: str
-
-        """
-        return self.poller.__class__.__name__
-
-    def remove_timeout(self, timeout_id):
-        """Remove a timeout if it's still in the timeout stack of the poller
-
-        :param str timeout_id: The timeout id to remove
-
-        """
-        self.poller.remove_timeout(timeout_id)
-
-    def start(self):
-        """Start the IOLoop, waiting for a Poller to take over."""
-        LOGGER.debug('Starting IOLoop')
-        self.poller.open = True
-        while not self.poller:
-            time.sleep(SelectPoller.TIMEOUT)
-        self.poller.start()
-        self.poller.flush_pending_timeouts()
-
-    def start_poller(self, handler, events, fileno):
-        """Start the Poller, once started will take over for IOLoop.start()
-
-        :param method handler: The method to call to handle events
-        :param int events: The events to handle
-        :param int fileno: The file descriptor to poll for
-
-        """
-        LOGGER.debug('Starting the Poller')
-        self.poller = None
+    def _get_poller(self):
+        """Determine the best poller to use for this enviroment."""
+    
         if hasattr(select, 'epoll'):
             if not SELECT_TYPE or SELECT_TYPE == 'epoll':
                 LOGGER.debug('Using EPollPoller')
-                self.poller = EPollPoller(fileno, handler, events,
-                                          self._manage_event_state)
-        if not self.poller and hasattr(select, 'kqueue'):
+                poller = EPollPoller()
+    
+        if not poller and hasattr(select, 'kqueue'):
             if not SELECT_TYPE or SELECT_TYPE == 'kqueue':
                 LOGGER.debug('Using KQueuePoller')
-                self.poller = KQueuePoller(fileno, handler, events,
-                                           self._manage_event_state)
-        if not self.poller and hasattr(select, 'poll') and hasattr(
-            select.poll(), 'modify'):
+                poller = KQueuePoller()
+        
+        if not poller and hasattr(select, 'poll') and hasattr(select.poll(),
+                                                              'modify'):
             if not SELECT_TYPE or SELECT_TYPE == 'poll':
                 LOGGER.debug('Using PollPoller')
-                self.poller = PollPoller(fileno, handler, events,
-                                         self._manage_event_state)
-        if not self.poller:
+                poller = PollPoller()
+        
+        if not poller:
             LOGGER.debug('Using SelectPoller')
-            self.poller = SelectPoller(fileno, handler, events,
-                                       self._manage_event_state)
-
-    def stop(self):
```

Comment:
@wjps, this triggers a bug in `BaseConnection._handle_ioloop_stop()` when it attempts to all `self.ioloop.stop()` here https://github.com/pika/pika/commit/e03c7adbb609f91e7523057a33c05817f096acad#diff-1a5308d8b684fdafac1f89e376e8933dR209

If confirmed, then in addition to the fix, it would be great to have a test for this code path.